Source/Proxy/IPublicCallContext.cs

```diff
@@ -0,0 +1,21 @@
+using System.Reflection;
+
+namespace Moq.Proxy
+{
+	/// <summary>
+	/// Class that represents the method call currently being executed. 
+	/// Allows for custom interception for open generic methods
+	/// </summary>
+	public interface IPublicCallContext
+	{
+		/// <summary>
+		/// Arguments passed into the method
+		/// </summary>
+		object[] Arguments { get; }
+
+		/// <summary>
+		/// Closed generic method currently being executed.
+		/// </summary>
+		MethodInfo Method { get; }
+	}
+}
```

Comment:
The reason I split it out, is to keep the `IPublicCallContext` as minimal as I can, without exposing some of the internal helper properties/methods which seemed to be `Castle.DynamicProxy` wrappers such as the following:

```
// Obsolete as the retunrr of the Func<T> is used, inline with rest of Moq
object ReturnValue { get; set; } 

// Could be potentially exposed? But rest of Moq uses a `.CallBase()` method?
void InvokeBase(); 

// Helper method Used internally by Moq?
void SetArgumentValue(int index, object value); 
```

`IPublicCallContext` is exposed in order to be able to construct meaningfull `.Returns()` calls as without it, it would be impossible to know what parameters the method was called with, nor the generic type parameters for the specific call.

It could be moved into the main `Moq` namespace as I guess `Moq.Proxy` seems to be `Castle.DynamicProxy` related items?